python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -542,42 +584,340 @@ def asset_key_for_input(self, input_name: str) -> AssetKey:
+                will be automatically provided.
+
+        Examples:
+            .. code-block:: python
+
+                partitions_def = DailyPartitionsDefinition("2023-08-20")
+
+                @asset(
+                    partitions_def=partitions_def
+                )
+                def an_asset(context: AssetExecutionContext):
+                    context.log.info(context.asset_partition_key_for_output())
+
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-21"
+
+                @multi_asset(
+                    outs={
+                        "first_asset": AssetOut(key=["my_assets", "first_asset"]),
+                        "second_asset": AssetOut(key=["my_assets", "second_asset"])
+                    }
+                    partitions_def=partitions_def,
+                )
+                def a_multi_asset(context: AssetExecutionContext):
+                    context.log.info(context.asset_partition_key_for_output("first_asset"))
+                    context.log.info(context.asset_partition_key_for_output("second_asset"))
+
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-21"
+                #   "2023-08-21"
+
+
+                @asset(
+                    partitions_def=partitions_def,
+                    ins={
+                        "self_dependent_asset": AssetIn(partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1))
+                    }
+                )
+                def self_dependent_asset(context: AssetExecutionContext, self_dependent_asset):
+                    context.log.info(context.asset_partition_key_for_output())
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-21"
+
         """
         return self._step_execution_context.asset_partition_key_for_output(output_name)
 
     @public
     def asset_partitions_time_window_for_output(self, output_name: str = "result") -> TimeWindow:
         """The time window for the partitions of the output asset.
 
+        If you want to write your asset to support running a backfill of several partitions in a single run,
+        you can use ``asset_partitions_time_window_for_output`` to get the TimeWindow of all of the partitions
+        being materialized by the backfill.
+
         Raises an error if either of the following are true:
         - The output asset has no partitioning.
         - The output asset is not partitioned with a TimeWindowPartitionsDefinition or a
         MultiPartitionsDefinition with one time-partitioned dimension.
+
+        Args:
+            output_name (str): The name of the output produced by the asset to get the time window for.
+                If not provided for assets defined with the `@asset` decorator, the name of the output
+                will be automatically provided.
```

Comment:
Editorial note: I think a longer term solution to the mental gymnastics required in all these descriptions will be to force the user to compose a call to map output name to asset key and then have all context methods be in terms of asset keys exclusively

```python
cxt.asset_partitions_time_window(cxt.asset_key_of_output("output_name"))
```

This would make things more simple at the expense of making things easy in the i/o manager case.

```python
ext.asset_partitions_time_window() # single asset case
ext.asset_partitions_time_window(asset_key) # multi asset case
```

Reply:
yeah i agree - basically what the deprecated methods became in this PR https://github.com/dagster-io/dagster/pull/16255/files#diff-890fa9a676b3b1f25888dcd0c569de82b6c56e3822b969b22c7a923c324fbcbbR913

Reply:
Quick update on this thread: I prototyped https://github.com/dagster-io/dagster/pull/16417 as a more fundamental fix for this problem